uitools/toolkitcpp/cpp/Esri/ArcGISRuntime/Toolkit/FieldsPopupElementViewController.cpp

```diff
@@ -0,0 +1,99 @@
+/*******************************************************************************
+ *  Copyright 2012-2024 Esri
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ ******************************************************************************/
+#include "FieldsPopupElementViewController.h"
+#include "PopupField.h"
+
+namespace Esri::ArcGISRuntime::Toolkit {
+
+/*!
+  \class Esri::ArcGISRuntime::Toolkit::FieldsPopupElementViewController
+  \inmodule ArcGISRuntimeToolkit
+  \ingroup ArcGISQtToolkitUiCppControllers
+  \brief In MVC architecture, this is the controller for the corresponding
+  \c FieldsPopupElementView.
+ */
+
+/*!
+ \brief Constructor
+ \list
+   \li \a parent Parent owning \c QObject.
+ \endlist
+ */
+FieldsPopupElementViewController::FieldsPopupElementViewController(QObject* parent)
+  : PopupElementViewItem{parent}
+{
+}
+
+/*!
+  \brief Destructor.
+  */
+FieldsPopupElementViewController::~FieldsPopupElementViewController() = default;
+
+/*!
+\brief Constructor. Takes a \a fieldsPopupElement and \a parent object.
+  */
+FieldsPopupElementViewController::FieldsPopupElementViewController(
+    QPointer<FieldsPopupElement> fieldsPopupElement, QObject* parent)
+  : PopupElementViewItem{QPointer<FieldsPopupElement>(std::move(fieldsPopupElement)), parent}
+{
+}
+
+/*!
+  \brief Returns the title of the \c FieldsPopupElement.
+ */
+QString FieldsPopupElementViewController::title() const
+{
+  return static_cast<FieldsPopupElement*>(popupElement())->title();
+}
+
+/*!
+  \brief Returns the combination of each element from \l {Esri::ArcGISRuntime::FieldsPopupElement::labels} {labels} and
+  \l {Esri::ArcGISRuntime::FieldsPopupElement::formattedValues} {formattedValues} of the
+  \c FieldsPopupElement in a QList of QMap elements
+  */
+QVariantList FieldsPopupElementViewController::values()
+{
+  const auto list1 = static_cast<FieldsPopupElement*>(popupElement())->labels();
+  const auto list2 = static_cast<FieldsPopupElement*>(popupElement())->formattedValues();
+  QVariantList combinedData;
+  int size = qMin(list1.size(), list2.size());
+
+  std::transform(list1.begin(), list1.begin() + size, list2.begin(),
+                 std::back_inserter(combinedData),
+                 [](const QString &item1, const QString &item2) {
+    QVariantMap item;
+    item["label"] = item1;
+    item["formattedValue"] = item2;
+    return item;
+  });
+  emit fieldsPopupElementChanged();
+  return combinedData;
+}
```

Comment:
I'm trying to wrap my head around this. It looks like you are iterating over both list1 and list2 at the same time. What happens if list1 and list2 have a different number of elements? Can that even happen?

Reply:
@JamesMBallard I don't believe it's possible that they have different number of elements. I think it's possible the fields related formatted value might be empty or null if you didn't give it a value but the lists should always be the same.

Reply:
In that case the `int size = qMin(list1.size(), list2.size());` is confusing to me if we can assume they're both the same size. I assume that is a safety check so we don't iterate beyond the end of either collection, but it leads the reader to believe they could be different sizes, which is what lead me to ask questions in the first place.

To me that means:
1. `int size = qMin(list1.size(), list2.size());` is unnecessary, or
2. We might leave off some elements if they're indeed different sizes.

My recommendation is to remove the `qMin` and use a Qt assert if the collections are different sizes. That should make it clearer that we are assuming they're the same size.

Reply:
Ah I see where I caused confusion now. Thanks for pointing this out.